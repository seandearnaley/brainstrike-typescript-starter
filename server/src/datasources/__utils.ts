// encodeCursor and decodeCursor code from https://github.com/kirkbrauer/typeorm-graphql-pagination by Kirk Brauer.
import { encode, decode } from "opaqueid";
import { Driver } from "typeorm";

export interface Edge<T> {
  cursor: string;
  node: T;
}

export interface PageInfoInterface {
  startCursor: string | null;
  endCursor: string | null;
  totalCount: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

/**
 * A cursor object.
 */
export interface Cursor {
  /**
   * The ID of the entity.
   */
  id: string;
  /**
   * The entity type.
   */
  type: string;
  /**
   * The entity index in the results. (we're using Row Number)
   */
  index: number;
}

/**
 * The invalid cursor error.
 */
export class InvalidCursorError extends Error {
  /**
   * Constructs a new InvalidCursorError.
   */
  constructor() {
    super();
    this.name = "Invalid Cursor Error";
    this.message = "Invalid cursor";
  }
}

/**
 * The invalid cursor type error.
 */
export class InvalidCursorTypeError extends Error {
  /**
   * The expected cursor type.
   */
  private expectedType: string;

  /**
   * The actual cursor type.
   */
  private actualType: string;

  /**
   * Constructs a new InvalidCursorTypeError
   * @param expectedType The expected cursor type.
   * @param actualType The actual cursor type.
   */
  constructor(expectedType: string, actualType: string) {
    super();
    this.name = "Invalid Cursor Type Error";
    this.expectedType = expectedType;
    this.actualType = actualType;
    this.message = `Invalid cursor, expected type ${expectedType}, but got type ${actualType}`;
  }
}

/**
 * Encodes a pagination cursor.
 * @param id The entity ID.
 * @param type The entity type.
 * @param index The entity index in the results.
 */
export function encodeCursor(id: string, type: string, index: number): string {
  return encode(`C|${type}|${id}|${index}`);
}

/**
 * Decodes a pagination cursor.
 * @param cursor The cursor to decode.
 * @param type The entity type.
 */
export function decodeCursor(cursor: string, type: string): Cursor {
  // Split the cursor
  const split: string[] = decode(cursor).split("|");
  // Verify that it is a valid cursor
  if (split[0] !== "C") throw new InvalidCursorError();
  // Throw an error if the cursor type is incorrect
  if (split[1] !== type) throw new InvalidCursorTypeError(type, split[1]);
  // Return the cursor data
  return {
    id: split[2],
    type: split[1],
    index: Number(split[3]) * 1,
  };
}

export function buildPageInfo<T extends { cursor: string }>(
  edges: T[],
  totalCount: number,
  type: string,
): PageInfoInterface {
  if (edges.length === 0) {
    // short circuit
    return {
      startCursor: null,
      endCursor: null,
      totalCount,
      hasNextPage: false,
      hasPreviousPage: false,
    };
  }

  const firstEdge = edges[0];
  const lastEdge = edges[edges.length - 1];

  const startCursor = firstEdge?.cursor ?? null;
  const endCursor = lastEdge?.cursor ?? null;

  if (!startCursor || !endCursor) throw Error("missing cursors");

  const lastRowNumber = Number(decodeCursor(endCursor, type).index);
  const firstRowNumber = Number(decodeCursor(startCursor, type).index);

  const hasNextPage = lastEdge ? lastRowNumber < totalCount : false;
  const hasPreviousPage = firstRowNumber > 1;

  return {
    startCursor,
    endCursor,
    totalCount,
    hasNextPage,
    hasPreviousPage,
  };
}

export function encodeGlobalID(id: string, __typename: string): string {
  return encode(`${id}:${__typename}`);
}

export function decodeGlobalID(objectId: string): {
  id: string;
  __typename: string;
} {
  const parts = decode(objectId).split(":");
  return {
    id: parts[0],
    __typename: parts[1],
  };
}

export function escapeStringsWithDriver(
  driver: Driver,
  ...strs: string[]
): string[] {
  return strs.map((str) => driver.escape(str));
}

export const getTypeName = (id: string): string => {
  return decodeGlobalID(id).__typename;
};
