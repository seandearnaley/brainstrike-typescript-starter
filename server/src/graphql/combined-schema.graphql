scalar Date
scalar DateTime
scalar Time

enum DirectionEnum {
  ASC
  DESC
}

interface Node {
  id: ID!
  created: DateTime!
  updated: DateTime
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
  endCursor: String
  totalCount: Int!
}

type Card implements Node {
  id: ID!
  number: Int
  label: String
  description: String
  created: DateTime!
  updated: DateTime
  categories: [Category!]!
}

type CardConnection {
  pageInfo: PageInfo!
  edges: [CardEdge!]!
}

type CardEdge {
  cursor: String!
  node: Card!
}

input CardInput {
  number: Int
  label: String
  description: String
  categoryId: ID
}

type CardsUpdatedResponse {
  success: Boolean!
  message: String!
  card: Card!
}

type Category implements Node {
  id: ID!
  name: String!
  updated: DateTime
  created: DateTime!
  cards(
    first: Int
    last: Int
    after: String
    before: String
    orderByColumn: String
    orderByDirection: DirectionEnum
  ): CardConnection!
  parent: Category
  children: [Category!]!
}

input CategoryInput {
  name: String!
  parentId: ID
}

type CategoryUpdatedResponse {
  success: Boolean!
  message: String!
  category: Category!
}

type Query {
  cards(
    first: Int
    last: Int
    after: String
    before: String
    orderByColumn: String
    orderByDirection: DirectionEnum
    categoryId: ID
  ): CardConnection!
  card(id: ID): Card!
  categories(
    cardIds: String
    orderByColumn: String
    orderByDirection: DirectionEnum
  ): [Category!]!
  category(id: ID): Category!
  node(id: ID): Node!
}

type Mutation {
  # card mutations
  addCard(input: CardInput): CardsUpdatedResponse!
  updateCard(id: ID!, input: CardInput): CardsUpdatedResponse!
  removeCard(id: ID!): CardsUpdatedResponse!
  # category mutations
  addCategory(input: CategoryInput): CategoryUpdatedResponse!
  updateCategory(id: ID!, input: CategoryInput): CategoryUpdatedResponse!
  removeCategory(id: ID!): CategoryUpdatedResponse!
}
