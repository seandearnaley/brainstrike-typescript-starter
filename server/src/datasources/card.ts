import DataLoader = require("dataloader");
import { DataSource, DataSourceConfig } from "apollo-datasource";
import { Connection } from "typeorm";
import { Card as CardEntity, Category as CategoryEntity } from "../entity";
import { CardInput, DirectionEnum } from "../generated/graphql";
import { ApolloContext } from "../types/context";
import { DataSourceRepos } from "../";

import {
  encodeCursor,
  decodeCursor,
  buildPageInfo,
  Edge,
  escapeStringsWithDriver,
  encodeGlobalID,
  decodeGlobalID,
} from "./__utils";

export type GetCardsArguments = {
  before?: string;
  after?: string;
  first?: number;
  last?: number;
  orderByColumn?: string;
  orderByDirection?: DirectionEnum;
  categoryId?: string;
};

// NOTE: using Omit here because categories is used on the Card entity to join to the categories table,
// but we actually want to use a dataloader for a "virtual" inner field using the same name "categories"
// we may want an even narrower type definition to give back to the resolver that does not overlap with
// the Card entity.
interface CardObject extends Omit<CardEntity, "categories"> {
  rowNumber?: number;
  categories: null;
}

type CardNodeObject = {
  cursor: string;
  node: CardObject;
};

type CardConnectionObject = {
  pageInfo: {
    hasNextPage: boolean;
    hasPreviousPage: boolean;
    startCursor: string;
    endCursor: string;
    totalCount: number;
  };
  edges: CardNodeObject[];
};

type CategoryLoader = {
  categoryId?: string;
  args?: GetCardsArguments;
};

type CardsUpdatedResponseObject = {
  success: boolean;
  message: string;
  card: CardObject;
};

export class CardAPI extends DataSource {
  context!: ApolloContext;
  connection: Partial<Connection>;
  repos: DataSourceRepos;

  constructor({ connection }: { connection: Partial<Connection> }) {
    super();
    this.connection = connection;
    this.repos = {
      cards: connection.getRepository(CardEntity),
      categories: connection.getRepository(CategoryEntity),
    };
  }

  /**
   * Apollo init function, called by apollo when setup
   * @param config used by apollo internally
   */
  initialize(config: DataSourceConfig<ApolloContext>): void {
    this.context = config.context;
  }

  /**
   * Get all cards in a deck, based on Relay Style Pagination Spec
   * https://facebook.github.io/relay/graphql/connections.htm#
   */
  async getCards({
    first = 100,
    last = 0,
    before = null,
    after = null,
    orderByColumn = "created",
    orderByDirection = DirectionEnum.Desc,
    categoryId,
  }: GetCardsArguments): Promise<CardConnectionObject> {
    // NOTE: we're using straight SQL for Postgres here because TypeORM has a bug where it can't resolve the subquery meta data,
    // it should be possible to rewrite this using the ORM syntax but will have to wait https://github.com/typeorm/typeorm/issues/4015

    // The advantage of this apporach using ROW_NUMBER () OVER is that it can be easily adapted to other engines, can do first and last,
    // is sortable, is executed in one query and needs no CTE's- if there is a better solution, let me know! seandearnaley@hotmail.com

    // escape input values for Postgres
    const [
      cursorColumn,
      cardTableName,
      categoryTableName,
      categoryCardsTableName,
      _orderByColumn,
    ] = escapeStringsWithDriver(
      this.connection.driver,
      "id",
      "card",
      "category",
      "category_cards_card",
      orderByColumn
    );

    // MANY-to-MANY JOIN
    const rowNumQuery = `
      SELECT ${cardTableName}.*, ROW_NUMBER () OVER (ORDER BY ${cardTableName}.${_orderByColumn} ${orderByDirection}) as "rowNumber" FROM ${cardTableName}
      ${
        categoryId // only need to add the joins if a categoryId is supplied, note its a many to many relationship, cards can be in multiple categories
          ? `
            LEFT JOIN ${categoryCardsTableName} on (${cardTableName}."id" = ${categoryCardsTableName}."cardId")
            LEFT JOIN ${categoryTableName} on (${categoryCardsTableName}."categoryId" = ${categoryTableName}."id")
            WHERE ${categoryCardsTableName}."categoryId" = '${
              decodeGlobalID(categoryId).id
            }'
          `
          : ""
      }
    `;

    const wheres = [];
    const params = [];

    if (after) {
      params.push(decodeCursor(after, "Card").id);

      wheres.push(
        `"rowNumber" > ( SELECT "t2"."rowNumber" FROM (${rowNumQuery}) as t2 WHERE ${cursorColumn} = $1 )`
      );
    }

    if (before) {
      params.push(decodeCursor(before, "Card").id);

      wheres.push(`
          "rowNumber" < (
            SELECT "t3"."rowNumber" FROM (${rowNumQuery}) as t3
            WHERE ${cursorColumn} = ${after ? "$2" : "$1"}
          )
        `);
    }

    if (first > 0) {
      if (after) {
        wheres.push(`
          "rowNumber" <= (
            SELECT "t4"."rowNumber" + ${first} FROM (${rowNumQuery}) as t4
            WHERE ${cursorColumn} = $1
          )
        `);
      } else {
        wheres.push(`"rowNumber" <= ${first}`);
      }
    }

    if (last > 0) {
      if (before) {
        wheres.push(`
          "rowNumber" >= (
            SELECT "t5"."rowNumber" - ${last} FROM (${rowNumQuery}) as t5
            WHERE ${cursorColumn} = ${after ? "$2" : "$1"}
          )
        `);
      } else {
        wheres.push(
          `"rowNumber" > ( SELECT MAX("t6"."rowNumber") - ${last} FROM (${rowNumQuery}) as t6 )`
        );
      }
    }

    let queryStr = `
      SELECT t1.* FROM (${rowNumQuery}) as t1
    `;

    if (wheres.length) queryStr += `WHERE ${wheres.join(" AND ")}`;

    const [results, [{ totalCount }]] = await Promise.all([
      this.connection.query(queryStr, [...params]),
      this.connection.query(
        `SELECT COUNT(*) as "totalCount" FROM (${rowNumQuery}) as countTable `
      ),
    ]);

    const edges = this.createEdges(results);
    const pageInfo = buildPageInfo<Edge<CardObject>>(
      edges,
      Number(totalCount),
      "Card"
    );

    return {
      edges: edges.map(({ node, cursor }) => ({
        node: this.encodeCard(node),
        cursor,
      })),
      pageInfo,
    };
  }

  protected encodeCard(data: CardEntity): CardObject {
    return {
      ...data,
      id: encodeGlobalID(data.id, "Card"), // replace ID with a global ID
      categories: null, // escaping cagegories here to replace them with virtual field
    };
  }

  protected createEdges(results: CardObject[]): Edge<CardObject>[] {
    return results.map((result: CardObject) => ({
      node: result,
      cursor: encodeCursor(result.id, "Card", result.rowNumber), // TODO: this cursor column could probably be dynamic
    }));
  }

  private cardLoader = new DataLoader<CategoryLoader, CardConnectionObject>(
    async (categoryIds: CategoryLoader[]): Promise<CardConnectionObject[]> =>
      Promise.all(
        categoryIds.map(async ({ categoryId, args }) =>
          // TODO: make cards somehow batch this into one query, this version will result in multiple getCards calls.
          // it's tricky to resolve because of how we're exploiting row number
          this.getCards({
            ...args,
            categoryId,
          })
        )
      )
  );

  async getCardConnectionFor(
    categoryId: string,
    args: GetCardsArguments
  ): Promise<CardConnectionObject> {
    return this.cardLoader.load({ categoryId, args });
  }

  /**
   * Get a particular card from the deck using global id
   * @param id global id
   */
  async getCardByGlobalID(id: string): Promise<CardEntity> {
    id = decodeGlobalID(id).id;
    const card = await this.repos.cards.findOne(id); // find by id

    if (!card) throw new Error("Card Not Found");
    return card; // NOTE: returning as a CardEntity rather a CardObject because it hasn't been encoded yet
  }

  /**
   * Get a particular card from the deck, returns encoded card
   * @param id global id
   */
  async getCard(id: string): Promise<CardObject> {
    const card = await this.getCardByGlobalID(id); // find by global id
    return this.encodeCard(card);
  }

  /**
   * Adds a new card to a deck
   * @param input card number, label, description
   */
  async addCard({
    number,
    label,
    description,
    categoryId,
  }: CardInput): Promise<CardsUpdatedResponseObject> {
    const card = new CardEntity();
    card.number = number;
    card.label = label;
    card.description = description;

    if (categoryId) {
      const category = await this.repos.categories.findOne(categoryId);
      card.categories = [category];
    } else {
      card.categories = null;
    }

    const savedCard = await this.repos.cards.save(card);
    return {
      success: true,
      message: "Card Added",
      card: this.encodeCard(savedCard),
    };
  }

  /**
   * Updates a card in a deck
   * @param id card uuid
   * @param input card number, label, description
   */
  async updateCard(
    id: string,
    { number, label, description, categoryId }: CardInput
  ): Promise<CardsUpdatedResponseObject> {
    const card = await this.getCardByGlobalID(id); // find by id
    card.number = number;
    card.label = label;
    card.description = description;

    if (categoryId) {
      const category = await this.repos.categories.findOne(categoryId);
      card.categories = [category];
    } else {
      card.categories = null;
    }

    const savedCard = await this.repos.cards.save(card);
    return {
      success: true,
      message: "Card Updated",
      card: this.encodeCard(savedCard),
    };
  }

  /**
   * Removes a card from the deck
   * @param id card uuid
   */
  async removeCard(id: string): Promise<CardsUpdatedResponseObject> {
    const card = await this.getCardByGlobalID(id); // find by id
    const originalCard = { ...card }; // remove wipes the ip, creating a copy for the card field
    await this.repos.cards.remove(card);
    return {
      success: true,
      message: "Card Removed",
      card: this.encodeCard(originalCard),
    };
  }
}
