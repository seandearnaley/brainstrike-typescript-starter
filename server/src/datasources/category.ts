import DataLoader = require("dataloader");
import { DataSource, DataSourceConfig } from "apollo-datasource";
import { Connection } from "typeorm";

import { Category as CategoryEntity } from "../entity";
import { CategoryInput, DirectionEnum } from "../generated/graphql";
import { ApolloContext } from "../types/context";
import { DataSourceRepos } from "..";
import { encodeGlobalID, decodeGlobalID } from "./__utils";

type GetCategoriesArguments = {
  orderByColumn?: string;
  orderByDirection?: DirectionEnum;
  cardIds?: string;
};

// NOTE: using Omit here because cards is used on the Category entity to join to the cards table,
// but we actually want to use a dataloader for a "virtual" inner field using the same name "cards"
// we may want an even narrower type definition to give back to the resolver that does not overlap with
// the Category entity.
interface CategoryObject
  extends Omit<CategoryEntity, "cards" | "parent" | "children"> {
  cards: null;
  parent: null;
  children: null;
}

type CategoryUpdatedResponseObject = {
  success: boolean;
  message: string;
  category: CategoryObject;
};

export class CategoryAPI extends DataSource {
  context!: ApolloContext;
  connection: Connection;
  repos: DataSourceRepos;

  constructor({ connection }: { connection: Connection }) {
    super();
    this.connection = connection;
    this.repos = {
      categories: connection.getRepository(CategoryEntity),
    };
  }

  /**
   * Apollo init function, called by apollo when setup
   * @param config used by apollo internally
   */
  initialize(config: DataSourceConfig<ApolloContext>): void {
    this.context = config.context;
  }

  protected encodeCategory(data: CategoryEntity): CategoryObject {
    const { id, name, created, updated } = data;
    return {
      id: encodeGlobalID(id, "Category"), // replace ID with a global ID
      name,
      created,
      updated,
      cards: null, // will use virtual inner queries for cards
      parent: null, // will use custom implementation for parent
      children: null, // will use custom implementation for children
    };
  }

  private categoryLoader = new DataLoader<string, CategoryObject[]>(
    async (cardIds: string[]): Promise<Array<CategoryObject[]>> => {
      // batches cardIds into a single call per request
      const categories = await this.getCategoryEntities({
        cardIds: cardIds.join(","),
      });

      return cardIds.map((id) =>
        categories
          .filter((category) =>
            category.cards.find((card) => card.id === decodeGlobalID(id).id)
          )
          .map(this.encodeCategory)
      );
    }
  );

  async getCategoriesFor(cardId: string): Promise<CategoryObject[]> {
    return this.categoryLoader.load(cardId);
  }

  /**
   * Get all categories, note we need the cards field on the entity here to filter on the dataloader
   * but we don't want to return that to the resolver because it will clash with our 'cards' virtual field
   */
  async getCategoryEntities({
    cardIds,
    orderByColumn = "category.name",
    orderByDirection = DirectionEnum.Asc,
  }: GetCategoriesArguments): Promise<CategoryEntity[]> {
    let query = this.repos.categories
      .createQueryBuilder("category")
      .leftJoinAndSelect("category.cards", "card");

    if (cardIds?.length) {
      // filter by CardID

      const decodedIds = cardIds
        .split(",")
        .map((encodedId) => decodeGlobalID(encodedId).id);

      query = query.where("card.id IN (:...cardIds)", { cardIds: decodedIds });
    }

    if (orderByColumn) {
      query = query.orderBy(orderByColumn, orderByDirection);
    }

    return query.getMany();
  }

  /**
   * Get all categories
   */
  async getCategories(args: GetCategoriesArguments): Promise<CategoryObject[]> {
    const data = await this.getCategoryEntities(args);

    return data.map(this.encodeCategory); // get all
  }

  /**
   * Get a particular category using global id
   * @param id global id
   */
  async getCategoryByGlobalID(id: string): Promise<CategoryEntity> {
    id = decodeGlobalID(id).id;
    const category = await this.repos.categories.findOne(id); // find by id

    if (!category) throw new Error("Category Not Found");
    return category;
  }

  /**
   * Get a particular category
   * @param id category uuid
   */
  async getCategory(id: string): Promise<CategoryObject> {
    const category = await this.getCategoryByGlobalID(id); // find by global id
    return this.encodeCategory(category);
  }

  /**
   * Adds a new category
   * @param input category name
   */
  async addCategory({
    name,
  }: CategoryInput): Promise<CategoryUpdatedResponseObject> {
    const category = new CategoryEntity();
    category.name = name;
    const savedCategory = await this.repos.categories.save(category);
    return {
      success: true,
      message: "Category Added",
      category: this.encodeCategory(savedCategory),
    };
  }

  /**
   * Updates a category
   * @param id category uuid
   * @param input category name
   */
  async updateCategory(
    id: string,
    { name }: CategoryInput
  ): Promise<CategoryUpdatedResponseObject> {
    const category = await this.getCategoryByGlobalID(id);
    category.name = name;
    const savedCategory = await this.repos.categories.save(category);
    return {
      success: true,
      message: "Category Updated",
      category: this.encodeCategory(savedCategory),
    };
  }

  /**
   * Removes a category from the deck
   * @param id category uuid
   */
  async removeCategory(id: string): Promise<CategoryUpdatedResponseObject> {
    const category = await this.getCategoryByGlobalID(id);
    const originalCategory = { ...category }; // remove wipes the ip, creating a copy for the category field

    id = decodeGlobalID(id).id;
    // remove dependant tree relations, unfortunately hasn't been implemented in TypeORM yet
    await this.repos.categories
      .createQueryBuilder()
      .delete()
      .from("category_closure") // check your db or migrations for the actual table name
      .where('"id_ancestor" = :id', { id })
      .execute();

    await this.repos.categories.remove(category);

    return {
      success: true,
      message: "Category Removed",
      category: this.encodeCategory(originalCategory),
    };
  }
}
