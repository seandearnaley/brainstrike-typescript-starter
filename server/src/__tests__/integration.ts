import {
  constructTestServer,
  createTestingConnection,
  DataSource,
  convertStringDatesToDateObjects,
} from "./__utils";

import * as TDATA from "./__testData";
import * as GQL from "./__queries";
import { CardAPI } from "../datasources/card";
import { print } from "graphql";

// Create type aliases for the non-exported types from CardAPI
type CardObject = Parameters<CardAPI["getCard"]>[0] extends string
  ? Awaited<ReturnType<CardAPI["getCard"]>>
  : never;

type CardConnectionObject = Awaited<ReturnType<CardAPI["getCards"]>>;

type CardsUpdatedResponseObject = Awaited<ReturnType<CardAPI["addCard"]>>;

describe("Queries", () => {
  let connection: DataSource;

  beforeAll(async () => {
    console.log("creating test connection");
    connection = await createTestingConnection();
  });

  afterAll(() => connection.close());

  it("fetches list of cards", async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({
        dataSources: {
          cardAPI,
          categoryAPI: {} as any,
        },
      }),
    });

    // mock the datasources' underlying fetch methods
    cardAPI.getCards = jest.fn(async () =>
      Promise.resolve(
        convertStringDatesToDateObjects(
          TDATA.mockCardsConnectionResult,
        ) as unknown as CardConnectionObject,
      ),
    );

    // Start the server - use try/catch to handle if already started
    try {
      await apolloServer.start();
    } catch (error) {
      // If the error is about the server already being started, we can ignore it
      if (!(error instanceof Error) || !error.message.includes("start()")) {
        throw error;
      }
    }

    // Execute the operation directly on the server
    const res = await apolloServer.executeOperation(
      {
        query: print(GQL.GET_CARD_DATA),
        variables: { first: 20 },
      },
      {
        contextValue: {
          dataSources: {
            cardAPI,
            categoryAPI: {} as any,
          },
        },
      },
    );

    expect(res).toMatchSnapshot();
  });

  it("fetches single card", async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({
        dataSources: {
          cardAPI,
          categoryAPI: {} as any,
        },
      }),
    });

    // mock the datasources' underlying fetch methods
    cardAPI.getCard = jest.fn(async () =>
      Promise.resolve(
        convertStringDatesToDateObjects(
          TDATA.mockFirstCardResponseEncoded,
        ) as unknown as CardObject,
      ),
    );

    // Start the server - use try/catch to handle if already started
    try {
      await apolloServer.start();
    } catch (error) {
      // If the error is about the server already being started, we can ignore it
      if (!(error instanceof Error) || !error.message.includes("start()")) {
        throw error;
      }
    }

    // Execute the operation directly on the server
    const res = await apolloServer.executeOperation(
      {
        query: print(GQL.GET_CARD),
        variables: { id: TDATA.mockFirstCardQueryId },
      },
      {
        contextValue: {
          dataSources: {
            cardAPI,
            categoryAPI: {} as any,
          },
        },
      },
    );

    expect(res).toMatchSnapshot();
  });
});

describe("Mutations", () => {
  let connection: DataSource;

  beforeAll(async () => {
    console.log("creating test connection");
    connection = await createTestingConnection();
  });

  afterAll(() => connection.close());

  it("create card", async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({
        dataSources: {
          cardAPI,
          categoryAPI: {} as any,
        },
      }),
    });

    // mock the datasources' underlying fetch methods
    cardAPI.addCard = jest.fn(async () =>
      Promise.resolve(
        convertStringDatesToDateObjects(
          TDATA.mockSuccessfulAddResponse,
        ) as unknown as CardsUpdatedResponseObject,
      ),
    );

    // Start the server - use try/catch to handle if already started
    try {
      await apolloServer.start();
    } catch (error) {
      // If the error is about the server already being started, we can ignore it
      if (!(error instanceof Error) || !error.message.includes("start()")) {
        throw error;
      }
    }

    // Execute the operation directly on the server
    const res = await apolloServer.executeOperation(
      {
        query: print(GQL.ADD_CARD),
        variables: { input: TDATA.mockCardInput },
      },
      {
        contextValue: {
          dataSources: {
            cardAPI,
            categoryAPI: {} as any,
          },
        },
      },
    );

    expect(res).toMatchSnapshot();
  });

  it("update card", async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({
        dataSources: {
          cardAPI,
          categoryAPI: {} as any,
        },
      }),
    });

    // mock the datasources' underlying fetch methods
    cardAPI.updateCard = jest.fn(async () =>
      Promise.resolve(
        convertStringDatesToDateObjects(
          TDATA.mockSuccessfulUpdateResponse,
        ) as unknown as CardsUpdatedResponseObject,
      ),
    );

    // Start the server - use try/catch to handle if already started
    try {
      await apolloServer.start();
    } catch (error) {
      // If the error is about the server already being started, we can ignore it
      if (!(error instanceof Error) || !error.message.includes("start()")) {
        throw error;
      }
    }

    // Execute the operation directly on the server
    const res = await apolloServer.executeOperation(
      {
        query: print(GQL.UPDATE_CARD),
        variables: {
          id: TDATA.mockFirstCardQueryId,
          input: TDATA.mockCardInput,
        },
      },
      {
        contextValue: {
          dataSources: {
            cardAPI,
            categoryAPI: {} as any,
          },
        },
      },
    );

    expect(res).toMatchSnapshot();
  });

  it("remove card", async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({
        dataSources: {
          cardAPI,
          categoryAPI: {} as any,
        },
      }),
    });

    // mock the datasources' underlying fetch methods
    cardAPI.removeCard = jest.fn(async () =>
      Promise.resolve(
        convertStringDatesToDateObjects(
          TDATA.mockSuccessfulRemoveResponse,
        ) as unknown as CardsUpdatedResponseObject,
      ),
    );

    // Start the server - use try/catch to handle if already started
    try {
      await apolloServer.start();
    } catch (error) {
      // If the error is about the server already being started, we can ignore it
      if (!(error instanceof Error) || !error.message.includes("start()")) {
        throw error;
      }
    }

    // Execute the operation directly on the server
    const res = await apolloServer.executeOperation(
      {
        query: print(GQL.REMOVE_CARD),
        variables: { id: TDATA.mockFirstCardQueryId },
      },
      {
        contextValue: {
          dataSources: {
            cardAPI,
            categoryAPI: {} as any,
          },
        },
      },
    );

    expect(res).toMatchSnapshot();
  });
});
