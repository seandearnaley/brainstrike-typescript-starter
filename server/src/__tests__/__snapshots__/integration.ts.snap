// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mutations > create card 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "addCard": {
          "card": {
            "__typename": "Card",
            "created": "2020-02-01T20:19:27.909Z",
            "description": "testing",
            "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
            "label": "Test Card 1",
            "number": 1,
            "updated": "2020-02-01T20:19:27.909Z",
          },
          "message": "Card Added",
          "success": true,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Mutations > remove card 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "removeCard": {
          "card": {
            "__typename": "Card",
            "created": "2020-02-01T20:19:27.909Z",
            "description": "testing",
            "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
            "label": "Test Card 1",
            "number": 1,
            "updated": "2020-02-01T20:19:27.909Z",
          },
          "message": "Card Removed",
          "success": true,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Mutations > update card 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "updateCard": {
          "card": {
            "__typename": "Card",
            "created": "2020-02-01T20:19:27.909Z",
            "description": "testing",
            "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
            "label": "Test Card 1",
            "number": 1,
            "updated": "2020-02-01T20:19:27.909Z",
          },
          "message": "Card Updated",
          "success": true,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Queries > fetches list of cards 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "cards": {
          "edges": [
            {
              "cursor": "Q3xDYXJkfDc1Zjc3NDczLWI3NGMtNGEyMC1hZTBkLTczZTA0YjQyN2NkZXwx",
              "node": {
                "__typename": "Card",
                "created": "2020-02-01T20:19:27.909Z",
                "description": "Amet autem et.",
                "id": "NzVmNzc0NzMtYjc0Yy00YTIwLWFlMGQtNzNlMDRiNDI3Y2RlOkNhcmQ=",
                "label": "sed possimus tempore",
                "number": 99999,
                "updated": "2020-02-01T20:19:27.909Z",
              },
            },
            {
              "cursor": "Q3xDYXJkfDc4NTU1YTkwLWY2OGItNDU0Yy1iZWIyLWU3MDIxZjNkOGM2M3wy",
              "node": {
                "__typename": "Card",
                "created": "2020-02-01T23:03:03.942Z",
                "description": "Odit officia esse hic qui est perspiciatis.",
                "id": "Nzg1NTVhOTAtZjY4Yi00NTRjLWJlYjItZTcwMjFmM2Q4YzYzOkNhcmQ=",
                "label": "necessitatibus id error",
                "number": 99995,
                "updated": "2020-02-01T23:03:03.942Z",
              },
            },
            {
              "cursor": "Q3xDYXJkfGNhZjM4YjMxLWU1NTEtNDMzYi1iMWIwLWM2NTgwN2U5NzkyN3wz",
              "node": {
                "__typename": "Card",
                "created": "2020-02-02T04:59:07.272Z",
                "description": "Et in sequi eum enim ratione.",
                "id": "Y2FmMzhiMzEtZTU1MS00MzNiLWIxYjAtYzY1ODA3ZTk3OTI3OkNhcmQ=",
                "label": "doloribus explicabo explicabo",
                "number": 99974,
                "updated": "2020-02-02T04:59:07.272Z",
              },
            },
          ],
          "pageInfo": {
            "endCursor": "Q3xDYXJkfGNhZjM4YjMxLWU1NTEtNDMzYi1iMWIwLWM2NTgwN2U5NzkyN3wz",
            "hasNextPage": true,
            "hasPreviousPage": false,
            "startCursor": "Q3xDYXJkfDc1Zjc3NDczLWI3NGMtNGEyMC1hZTBkLTczZTA0YjQyN2NkZXwx",
            "totalCount": 10000,
          },
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Queries > fetches single card 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "card": {
          "__typename": "Card",
          "created": "2020-02-01T20:19:27.909Z",
          "description": "testing",
          "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
          "label": "Test Card 1",
          "number": 1,
          "updated": "2020-02-01T20:19:27.909Z",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
